iceberg-rust/src/table/transaction/append.rs

```diff
@@ -0,0 +1,87 @@
+use std::cmp::Ordering;
+
+use iceberg_rust_spec::manifest::ManifestEntry;
+use smallvec::{smallvec, SmallVec};
+
+use crate::{
+    error::Error,
+    util::{cmp_with_priority, partition_struct_to_vec, try_sub, Rectangle},
+};
+
+/// Split sets of datafiles depending on their partition_values
+#[allow(clippy::type_complexity)]
+pub fn split_datafiles_once(
+    files: impl Iterator<Item = Result<ManifestEntry, Error>>,
+    rect: Rectangle,
+    names: &[&str],
+) -> Result<[(Vec<ManifestEntry>, Rectangle); 2], Error> {
+    let mut smaller = Vec::new();
+    let mut larger = Vec::new();
+    let mut smaller_rect = None;
+    let mut larger_rect = None;
+
+    for manifest_entry in files {
+        let manifest_entry = manifest_entry?;
+        let position = partition_struct_to_vec(manifest_entry.data_file().partition(), names)?;
+        // Compare distance to upper and lower bound. Since you can't compute a "norm" for a multidimensional vector where the dimensions have different datatypes,
+        // the dimensions are compared individually and the norm is computed by weighing the earlier columns more than the later.
+        if let Ordering::Greater = cmp_with_priority(
+            &try_sub(&position, &rect.min)?,
+            &try_sub(&rect.max, &position)?,
+        )? {
+            // if closer to upper bound
+            larger.push(manifest_entry);
+
+            if larger_rect.is_none() {
+                larger_rect = Some(Rectangle::new(position.clone(), position));
+            } else if let Some(larger_rect) = larger_rect.as_mut() {
+                larger_rect.expand_with_node(position);
+            }
+        } else {
+            // if closer to lower bound
+            smaller.push(manifest_entry);
+
+            if smaller_rect.is_none() {
+                smaller_rect = Some(Rectangle::new(position.clone(), position));
+            } else if let Some(smaller_rect) = smaller_rect.as_mut() {
+                smaller_rect.expand_with_node(position);
+            }
+        }
+    }
+    Ok([
+        (
+            smaller,
+            smaller_rect.ok_or(Error::NotFound("Lower".to_owned(), "rectangle".to_owned()))?,
+        ),
+        (
+            larger,
+            larger_rect.ok_or(Error::NotFound("Upper".to_owned(), "rectangle".to_owned()))?,
+        ),
+    ])
+}
+
+/// Splits the datafiles *n_split* times to decrease the number of datafiles per maniefst. 1 split returns 2 outputs vectors, 2 splits return 4, 3 splits return 8 and so on.
+pub(crate) fn split_datafiles(
+    files: impl Iterator<Item = Result<ManifestEntry, Error>>,
+    rect: Rectangle,
+    names: &[&str],
+    n_split: u32,
+) -> Result<SmallVec<[Vec<ManifestEntry>; 2]>, Error> {
```

Comment:
```suggestion
) -> Result<Vec<Vec<ManifestEntry>>, Error> {
```

The optimization is not necessary here as `split_datafiles` is called once per append, and using `Vec` is substantially more readable.